### ISSUE TITLE
__sync_add_and_fetch does not work?

### ISSUE THREAD

**Issue by craig65535** [iovisor/bcc#2654]
I'm using bcc release v0.11.0 (0fa419a64e71984d42f107c210d3d3f0cc82d59a) and RHEL 8.0 (kernel 4.18.0-80.11.2.el8_0.x86_64).

I noticed that `__sync_add_and_fetch` within my program has no meaningful return value (it always returns 2). It **does** increase the value of the variable passed in the first parameter, but doesn't return its new value.

Is this expected?

Here's a program to reproduce:

```C++
#include <iostream>

#include "bcc/BPF.h"

using namespace std;

static const string _probe_fork_counter = R"(
#include <uapi/linux/ptrace.h>

typedef struct event {
    uint64_t id1;
    uint64_t id2;
} event_t;

typedef struct counter {
    uint64_t id;
} counter_t;

BPF_ARRAY(counter_arr, counter_t, 1);

BPF_PERF_OUTPUT(events);

int do_fork_probe(struct pt_regs *ctx)
{
    event_t event = { 0 };
    counter_t *counter;
    int idx = 0;
    counter = counter_arr.lookup(&idx);
    if (counter) {
        /* atomic: */
        event.id1 = __sync_add_and_fetch(&counter->id, 1);
        /* NOT ATOMIC: */
        (void)__sync_add_and_fetch(&counter->id, 1);
        event.id2 = counter->id;
        events.perf_submit(ctx, &event, sizeof(event));
    }
    return 0;
}
)";

/* This definition must mirror the one in the BPF program above. */
typedef struct event {
    uint64_t id1;
    uint64_t id2;
} event_t;

static uint64_t my_counter;

static void _output(void *cb_ctx, void *data, int data_size)
{
    (void)cb_ctx;
    if (data_size < 0 || (size_t)data_size < sizeof(event_t)) {
        cerr << "data size mismatch (" << data_size << " < expected " << sizeof(event_t) << ")" << endl;
    } else {
        auto event = static_cast<event_t *>(data);
        cout << "id1: " << event->id1 << " id2: " << event->id2 << endl;
        /* atomic: */
        uint64_t my_counter1 = __sync_add_and_fetch(&my_counter, 1);
        /* NOT ATOMIC: */
        (void)__sync_add_and_fetch(&my_counter, 1);
        uint64_t my_counter2 = my_counter;
        cout << "my_counter1: " << my_counter1 << " my_counter2: " << my_counter2 << endl;
    }
}

int main()
{
    ebpf::BPF bpf;
    auto init_res = bpf.init(_probe_fork_counter);
    if (init_res.code() != 0) {
        throw runtime_error(init_res.msg()); 
    }

    auto attach_res = bpf.attach_kprobe("_do_fork", "do_fork_probe");
    if (attach_res.code() != 0) {
        throw runtime_error(attach_res.msg()); 
    }

    /* Set callback */
    auto open_res = bpf.open_perf_buffer("events", &_output);
    if (open_res.code() != 0) {
        throw runtime_error(open_res.msg()); 
    }

    cout << "Started" << endl;
    do {
        (void)bpf.poll_perf_buffer("events");
    } while (true);

    return 1;
}
```

Expected output:

> Started
id1: 1 id2: 2
my_counter1: 1 my_counter2: 2
id1: 3 id2: 4
my_counter1: 3 my_counter2: 4
id1: 5 id2: 6
my_counter1: 5 my_counter2: 6

Actual output:

> Started
id1: **2** id2: 2
my_counter1: 1 my_counter2: 2
id1: **2** id2: 4
my_counter1: 3 my_counter2: 4
id1: **2** id2: 6
my_counter1: 5 my_counter2: 6

**Comment by craig65535** [comment]
(I'm using LLVM/clang 7.0.1)

**Comment by yonghong-song** [comment]
Yes. This is expected. If you use a newer compiler (llvm 8.0.0 and later), you will get a fatal error to avoid such a surprise.

**Comment by craig65535** [comment]
I'll try a newer LLVM - thank you.

Incidentally are you aware of another way to accomplish something like this?

**Comment by yonghong-song** [comment]
If you worry about atomicity, you can try bpf_spin_lock. But it only available in newer kernels, 5.1.

**Comment by yonghong-song** [comment]
Please use llvm >= 12, __sync_add_and_fetch() should be supported properly. Let me know if it doesn't.
